KEYCLOAK-5244 Add BlacklistPasswordPolicyProvider

thomasdarimont opened the pull request:
This introduces a new `PasswordPolicy` which can refer to
a named predefined password-blacklist to avoid users
choosing too easy to guess passwords.

The `BlacklistPasswordPolicyProvider` supports built-in as
well as custom blacklists.
built-in blacklists use the form `default/filename`
and custom ones `custom/filename`, where filename
is the name of the found blacklist-filename.

I'd propose to use some of the freely available password blacklists
from the [SecLists](https://github.com/danielmiessler/SecLists/tree/master/Passwords) project.

Note: Due to licensing restrictions it might not be possible to include those lists.

For testing purposes one can download the password blacklist
```
wget -O 10_million_password_list_top_1000000.txt https://github.com/danielmiessler/SecLists/blob/master/Passwords/10_million_password_list_top_1000000.txt?raw=true
```
to /data/keycloak/blacklists/

Custom password policies can be configured with the SPI (See javadoc of `BlacklistPasswordPolicyProviderFactory` for details.
configuration mechanism via jboss-cli:
```
/subsystem=keycloak-server/spi=password-policy:add()
/subsystem=keycloak-server/spi=password-policy/provider=passwordBlacklist:add(enabled=true)
/subsystem=keycloak-server/spi=password-policy/provider=passwordBlacklist:write-attribute(name=properties.blacklistsFolderUri, value=file:///data/keycloak/blacklists/)
```

thomasdarimont committed:
KEYCLOAK-5244 Add BlacklistPasswordPolicyProvider

This introduces a new PasswordPolicy which can refer to
a named predefined password-blacklist to avoid users
choosing too easy to guess passwords.

The BlacklistPasswordPolicyProvider supports built-in as
well as custom blacklists.
built-in blacklists use the form `default/filename`
and custom ones `custom/filename`, where filename
is the name of the found blacklist-filename.

I'd propose to use some of the freely available password blacklists
from the [SecLists](https://github.com/danielmiessler/SecLists/tree/master/Passwords) project.

For testing purposes one can download the password blacklist
```
wget -O 10_million_password_list_top_1000000.txt https://github.com/danielmiessler/SecLists/blob/master/Passwords/10_million_password_list_top_1000000.txt?raw=true
```
to /data/keycloak/blacklists/

Custom password policies can be configured with the SPI
configuration mechanism via jboss-cli:
```
/subsystem=keycloak-server/spi=password-policy:add()
/subsystem=keycloak-server/spi=password-policy/provider=passwordBlacklist:add(enabled=true)
/subsystem=keycloak-server/spi=password-policy/provider=passwordBlacklist:write-attribute(name=properties.blacklistsFolderUri, value=file:///data/keycloak/blacklists/)
```

Password blacklist is stored in a TreeSet.

hmlnarik commented on pom.xml:
```diff
@@ -88,6 +88,7 @@
         <version.jboss-integration-platform>6.0.10.Final</version.jboss-integration-platform>
 
         <!-- Others -->
+        <commons.collections4.version>4.1</commons.collections4.version>
```
Apache commons collections 4 are not available in wildfly, this would need a new dependency which is not a trivial task. I think the only reason for including it is using Radix (Patricia) tree.

hmlnarik reviewed (commented):
Why was Radix tree chosen over standard Set implementations?

thomasdarimont commented:
Hello @hmlnarik, Hello @pedroigor, 

I just revised the implementation again. I now use a plain `TreeSet` as a backing store for the blacklist. This should be good enough for medium size blacklists, e.g. up to 10k entries.

I plan to provide a separate blacklist password policy implementation with support for 
larger password blacklists, with up to 10e7 entries, as an custom extension module later.

Is the implementation now okay?

stianst commented:
Is the reason for a separate custom extension the need for dependencies to do large collections? Is there any libs already included in WildFly that could be used or another way to do it if so?

thomasdarimont commented:
Hi Stian thanks for having a look at this.

I used a trie data structure to efficiently deal with large blacklists (I
used 1e6 and 1e7) entries. I couldn't find a suitable comparble
datastructure in the libraries available in current wildfly.

I removed the commons-collections4 dependency in one of the latest commits
so this shouldn't be an issue anymore. Plain Java collections are ok for
smallish blacklists (imho).

Nevertheless it might be worthwhile to investigate more appropriate
datastructures of one want to support really large blacklists. On the
mailing list I mentioned using Bloomfilters for this.

Stian Thorgersen <notifications@github.com> schrieb am Di., 15. Aug. 2017,
09:29:

> Is the reason for a separate custom extension the need for dependencies to
> do large collections? Is there any libs already included in WildFly that
> could be used or another way to do it if so?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/pull/4370#issuecomment-322396341>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATNQpjR5IfoMK6Tk045nCSGM5V6_tgSks5sYUi7gaJpZM4OsPw4>
> .
>

stianst commented:
A single implementation that can handle small and bigger dictionaries would be best IMO. What about Guava? It has BloomFilter (http://google.github.io/guava/releases/19.0-rc2/api/docs/com/google/common/hash/BloomFilter.html).

thomasdarimont commented:
it seems that the latest wildfly ships with guava 20.0 which I could use.
This means that I'd need to add a guava dependency to `keycloak-parent` and
`keycloak-server-spi-private`, would this be ok?

thomasdarimont committed:
KEYCLOAK-5244 Encode PasswordBlacklist as a BloomFilter

We now use a dynamically sized BloomFilter with a
false positive probability of 1% as a backing store
for PasswordBlacklists.

BloomFilter implementation is provided by google-guava
which is available in wildfly.

Password blacklist files are now resolved against
the ${jboss.server.data.dir}/password-blacklists.

This can be overridden via system property, or SPI config.
See JavaDoc of BlacklistPasswordPolicyProviderFactory for details.

Revised implementation to be more extensible, e.g. it could be
possible to use other stores like databases etc.

Moved FileSystem specific methods to FileBasesPasswordBlacklistPolicy.

The PasswordBlacklistProvider uses the guava version 20.0
shipped with wildfly. Unfortunately the arquillian testsuite
transitively depends on guava 23.0 via the selenium-3.5.1
dependency. Hence we need to use version 23.0 for tests but 20.0
for the policy provider to avoid NoClassDefFoundErrors in the
server-dist.

Configure password blacklist folder for tests

thomasdarimont commented:
@stianst I updated the PR with the BloomFilter implementation.

Instead of having some default blacklists for testing purposes in the jar, I'd rather put the test files somewhere under src/test/resources, wdyt?

I'm also not to excited about the blacklist file handling, e.g. whether the resource comes from a folder in the classpath, from a jar or a file system location.
My current approach is to only scan certain locations for blacklists in order to avoid  to expose filesystem folders.

hmlnarik commented:
@thomasdarimont Would it be reasonable to provide let say `PasswordBlacklistWordlistProvider` SPI that would only be used for providing the passwords? Then you could have `PasswordBlacklistProvider` load the list directly from all `PasswordBlacklistWordlistProvider`s in the system which would:
* Make testing easy - you would add the provider to [testsuite providers](https://github.com/keycloak/keycloak/tree/master/testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers)
* The same provider would set an example on loading static lists via your provider from classpath
* Make it easy to obtain the lists from web (e.g. using another provider)
* Add custom locations as needed in the future

thomasdarimont commented:
Hi  Hynek,

thought about that too.
https://github.com/keycloak/keycloak/pull/4370/files#diff-fb3ffc957154b02c22d09621e9343712R104

 I'm going to introduce another spi then.

Cheers,
Thomas

Am 17.08.2017 7:41 vorm. schrieb "Hynek Mlnařík" <notifications@github.com>:

> @thomasdarimont <https://github.com/thomasdarimont> Would it be
> reasonable to provide let say PasswordBlacklistWordlistProvider SPI that
> would only be used for providing the passwords? Then you could have
> PasswordBlacklistProvider load the list directly from all
> PasswordBlacklistWordlistProviders in the system which would:
>
>    - Make testing easy - you would add the provider to testsuite providers
>    <https://github.com/keycloak/keycloak/tree/master/testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers>
>    - The same provider would set an example on loading static lists via
>    your provider from classpath
>    - Make it easy to obtain the lists from web (e.g. using another
>    provider)
>    - Add custom locations as needed in the future
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/pull/4370#issuecomment-322974107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATNQlB40DVqu_3jmAjYPLLxnDrZ_-f6ks5sY9KKgaJpZM4OsPw4>
> .
>

stianst commented:
-1 To PasswordBlacklistWordlistProvider SPI. That really is overkill. Rather just make it easy to extend this provider.

stianst commented:
I would just have it load the file by filename only from a specific directory within the server (data dir).

thomasdarimont commented:
Okay sounds better. Will adapt the PR then.

Cheers,
Thomas

Am 17.08.2017 8:35 vorm. schrieb "Stian Thorgersen" <
notifications@github.com>:

> I would just have it load the file by filename only from a specific
> directory within the server (data dir).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/pull/4370#issuecomment-322981954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATNQrwN0v7HUm0RT9knIoFMiuAHHECPks5sY99HgaJpZM4OsPw4>
> .
>

thomasdarimont commented:
@stianst 
I reworked the `BlacklistPasswordPolicyProvider`and friends to find password-blacklists 
beneath a single folder.

This folder is determined as:
- the value of the System property `keycloak.password.blacklists.path` if configured (fails if folder is missing)
- the value of the SPI config property: blacklistsPath when explicitly configured (fails if folder is missing)
- otherwise ${jboss.server.data.dir}/password-blacklists/ if nothing else is configured (the folder is created automatically)

In order to refer to a password blacklist-file users only have to enter the filename that lives in the configured password blacklist-folder.

Problem 1:
The problem I'm now facing is that I get a `NoClassDefinitionFoundException` at runtime 
which says that a `com.google.guava` class cannot be found.

How can I get the guava dependency added to `module.xml` of `keycloak-server-spi-private`?
Where is the `keycloak-server-spi-private` module configured?

Tried to define a guava dependency in `org.keycloak.subsystem.server.extension.KeycloakProviderDependencyProcessor`
but it's configuration is skipped for `keycloak-server` modules.
(Btw. it would be neat if guava would be available to custom providers by default...)

If I manually edit `module.xml` in
`$KEYCLOAK_HOME/modules/system/layers/keycloak/org/keycloak/keycloak-server-spi-private/main/module.xml`

by adding
```
<dependencies>
  ...
  <module name="com.google.guava"/>
</dependencies>
```
then it works.


Problem 2:
How can I refer to the files under `keycloak/testsuite/integration-arquillian/tests/base/src/test/resources/password-blacklists`? 
How can I add a SystemProperty like:
```
-Dkeycloak.password.blacklists.path=/home/tom/dev/repos/gh/thomasdarimont/keycloak-dev/keycloak/testsuite/integration-arquillian/tests/base/src/test/resources/password-blacklists```
be applied to the Arquillian instance which runs the `PasswordPolicyTest`?

stianst commented:
Sorry for the late response. 

Problem 1:
You can add dependencies to guave in https://github.com/keycloak/keycloak/blob/master/distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-spi-private/main/module.xml

Problem 2:
Why not simply set the sys prop in a @before in the test itself and clear it in the @after?

Problem 3:
Tests need to work both on embedded Undertow (just run the tests from IDE) and the proper dist (Travis does this, but you can also do it yourself with -Pauth-server-wildfly, more details in the testsuite README).

thomasdarimont commented:
Thanks for the feedback!

Taking a look at this right now.

thomasdarimont commented:
ad Problem 1:
This seems to be solved, thanks :)

ad Problem 2:
Been there, done that...

neither adding the following in `PasswordPolicyTest`
```
    @Before
    public void setup() throws Exception{

//        // TODO how to set this system property in the spawed arquillian instance?
        File blacklistsLocation = new File(PasswordPolicyTest.class.getClassLoader().getResource("password-blacklists").toURI());
        System.setProperty("keycloak.password.blacklists.path", blacklistsLocation.getAbsolutePath());
    }
```
nor... overriding the `beforeAbstractKeycloakTest()` method:
```
    @Override
    public void beforeAbstractKeycloakTest() throws Exception {
        super.beforeAbstractKeycloakTest();

        File blacklistsLocation = new File(PasswordPolicyTest.class.getClassLoader().getResource("password-blacklists").toURI());
        System.setProperty("keycloak.password.blacklists.path", blacklistsLocation.getAbsolutePath());
    }
```
sets the System Property for the spawned arquillian process. How can I ensure the System Property is set in the spawned arquillian process in which the server run?

ad Problem 3:
Yes I know, will make sure that this works.

thomasdarimont commented:
@stianst seems that I need to provide a way to configure JVM arguments via `arquillian.xml` 
in `org.keycloak.testsuite.arquillian.undertow.KeycloakOnUndertowConfiguration`.

I need to set: 
```
<property name="javaVmArguments">
                -Dkeycloak.password.blacklists.path=password-blacklists
            </property>
```

mposolda commented:
@thomasdarimont There is also possibility to use custom REST endpoint for set system property and then remove it. We use it for example in TestingExportImportResource . Maybe it makes sense to add generic endpoint for set system property?

This has advantage, that you can change property during test and ensure that just this particular test will use this system property and not clash with other tests (Which I guess arquillian.xml based approach is problematic?) . Also there is possibility that you can use RunOnServer for set system property, but not 100% sure...

stianst commented:
+1 To Marek's suggestion. If you use RunOnServer to set the system prop it'll be available regardless how the server is ran (embedded Undertow or separate dist process).

thomasdarimont commented:
Sorry for the back and forth, but I tried that before - I gave yet it another spin, but couldn't get it to work.

I changed the test to:
```java
    /**
     * KEYCLOAK-5244
     */
    @Test
    public void testBlacklistPasswordPolicyWithTestBlacklist() throws Exception {

        final Path blackListsPath = Paths.get(Thread.currentThread().getContextClassLoader().getResource("password-blacklists").toURI()).toAbsolutePath();

        testingClient.server("passwordPolicy").run(session -> {

            try (SetSystemProperty ssp = new SetSystemProperty("keycloak.password.blacklists.path", blackListsPath.toString())) {

                RealmModel realmModel = session.getContext().getRealm();
                PasswordPolicyManagerProvider policyManager = session.getProvider(PasswordPolicyManagerProvider.class);

                realmModel.setPasswordPolicy(PasswordPolicy.parse(session, "passwordBlacklist(test-password-blacklist.txt)"));

                Assert.assertEquals(BlacklistPasswordPolicyProvider.ERROR_MESSAGE, policyManager.validate("jdoe", "blacklisted1").getMessage());
                Assert.assertEquals(BlacklistPasswordPolicyProvider.ERROR_MESSAGE, policyManager.validate("jdoe", "blacklisted2").getMessage());
                assertNull(policyManager.validate("jdoe", "notblacklisted"));
            }
        });
    }
```

Problem is that the `BlacklistPasswordPolicyProviderFactory` is initialized only
once for all test cases in `PasswordPolicyTest`, thus the first test-case that runs triggers the initialization of the `BlacklistPasswordPolicyProviderFactory` which then tries to determine the location of the blacklist folder.

Setting the System Property via the test REST interface would result in the same 
problem IMHO.

Even if I move the code to `beforeAbstractKeycloakTest()` the System Property is not set on the server when the `BlacklistPasswordPolicyProviderFactory` is initialized since `beforeAbstractKeycloakTest()` is executed *after* the provider initialization.

That's the reason why I ended up setting the System Property in the `pom.xml` as described above.

kkplein commented:
Nice to see work on this, it is appreciated, and followed with great interest by us. One question: would your provider also work to blacklist specific single characters? (like: , or &)

thomasdarimont commented:
@kkplein 
thanks for your interest in this, but this is not supported by this PasswordPolicy.
The idea of this Blacklist-PasswordPolicy is to not allow an explicitly given set 
of whole passwords.

However Keycloaks flexible password policies can be combined.
In your case you could use the Regex-PasswordPolicy (which is already built-in to Keycloak) 
to only allow passwords that don't contain `,` or `&`.

E.g. with the following regex:`^[^,&]+$`

kkplein commented:
ok! Thanks for the reply, and the regex suggestion. Looking forward to your plugin becoming mainstream, so we can start using it. :-)

stianst commented:
Setting the system property before running the tests is far from ideal and in the future we will want to be able to point the testsuite to an external Keycloak instance (running on OpenShift for instance).

What about using runOnServer to create the file in "${jboss.server.data.dir}/password-blacklists/"? Then delete once test is finished?

thomasdarimont commented:
Thanks for the suggestion, I give it a try.

thomasdarimont commented:
Unfortunately the system property `${jboss.server.data.dir}` is not defined during the test-suite run... 
other options?

stianst commented:
We can set the jboss.server.data.dir to something sensible in KeycloakServer and KeycloakOnUndertow. Maybe just user temp directory would do?

thomasdarimont committed:
KEYCLOAK-5244 Configure jboss.server.data.dir for test servers

thomasdarimont commented:
@stianst I rebased and adapted the PR as you suggested.

I squashed the commits to remove noise.

farmersmurf commented:
This is an import feature for me was just, wondering when this PR is likely to get the go ahead

kkplein commented:
Yes, same here. We're looking forward to this feature as well! :-)

stianst commented:
I will review this soon, but I also need to check with our QE team if we can merge it now. As we're past our feature freeze for 3.x we may need to mark this as a preview feature in RH-SSO or wait with merging until 4.x.

pdrozd commented:
@thomasdarimont Thanks for the PR, looks good to me. One minor thing I noticed:
Please add the invalidPasswordBlacklistedMessage  message also to themes/base/login/messages/messages_en.properties

thomasdarimont committed:
KEYCLOAK-5244 Translate blacklisted message in base/login

thomasdarimont commented:
@pdrozd thanks for review. Just added the english translation to base/login.

stianst commented:
Thanks @thomasdarimont

thomasdarimont commented:
Thanks for merging! :)

dfherr commented:
I know this is 2 years closed and merged and maybe I have a lack of understanding, but considering it was discussed here that small and larger password blacklists should be supported by a single implementation for both, why is this handled with a file that is loaded in memory?

To actually support large password blacklists (e.g. https://haveibeenpwned.com/Passwords) the passwords should be managed in a more structured way than loading them in memory from a file. Loading them into a database with an index (preferably a trie-based one) would provide the necessary lookup speed and be only limited by database disk size. Trie lookup itself is O(M) with M being the length of the password (hash). Would also solve the bluntness of falsely rejecting good passwords with a 1% probability.
